packages/data-table-server/src/middleware/attachDataTableFromPreviewToContext.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2022 Beyond Essential Systems Pty Ltd
+ */
+
+import { NextFunction, Request, Response } from 'express';
+import { getAjv } from '@tupaia/tsutils';
+import { generateId } from '@tupaia/database';
+import { DataTableSchema } from '@tupaia/types';
+import type { DataTable } from '@tupaia/types';
+
+import { DataTableServiceBuilder, getDataTableServiceType } from '../dataTableService';
+
+const validateDataTableFields = (dataTableFields: any): DataTable => {
+  const ajv = getAjv();
+  const dataTableValidator = ajv.compile(DataTableSchema);
+  // Add random id to meet validation requirement
+  const validate = dataTableValidator({ id: generateId(), ...dataTableFields });
+  if (!validate && dataTableValidator.errors) {
+    throw new Error(dataTableValidator.errors[0].message);
+  }
+
+  return dataTableFields;
+};
+
+/**
+ * Finds the requested dataTable and attaches it to the context
+ * Checks for permissions
+ */
+export const attachDataTableFromPreviewToContext = async (
+  req: Request<any, any, any, any>,
+  res: Response,
+  next: NextFunction,
+) => {
+  try {
+    const { accessPolicy, models, ctx, body } = req;
+    const dataTableFields = validateDataTableFields(body);
+
+    const dataTable = await models.dataTable.generateInstance(dataTableFields);
+    const { code: dataTableCode } = dataTable;
+    const { permission_groups: permissionGroups } = dataTable;
+    if (
```

Comment:
Can DRY a lot of this with the existing attachDataTableToContext function. I suggest extracting this out to `validatePermissions(dataTable, permissionGroups)` and `getDataTableService(dataTable)`